File: src/steps/step_create.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# IMPORTS
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

from src.local.misc import *;
from src.local.system import *;
from src.local.typing import *;

from src.core.log import *;
from src.core.utils import createFile;
from src.core.utils import createPath;
from src.core.utils import writeTextFile;
from src.customtypes.exports import ProjectTree;
from src.setup import appconfig;
from src.parsers.methods import convertToPythonString;

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# GLOBAL VARIABLES
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

#

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# METHOD: step create
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

def step():
    logInfo('CREATION STAGE STARTED.');
    root = appconfig.getPathRoot();
    createFilesAndFolders(path=root, projectTree=appconfig.getProjectTree());
    if appconfig.getWithFileStamp():
        createFileStamp(
            path=appconfig.getFileStamp(rel=False),
            overwrite=appconfig.getOptionOverwriteStamp(),
            options=appconfig.getDictionaryStamp()
        );
    if appconfig.getWithFileParamsPy():
        createParameters(options=appconfig.getDictionaryParms());
    logInfo('CREATION STAGE COMPLETE.');
    return;

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# SECONDARY METHODS
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

def createFilesAndFolders(path: str, projectTree: ProjectTree):
    for relpath in projectTree.directories():
        if not make_dir_if_not_exists(path=path, fname=relpath):
            raise FileExistsError('Could not create (sub)folder \033[1m{}\033[0m'.format(relpath));
    for relfname in projectTree.files():
        if not make_file_if_not_exists(path=path, fname=relfname):
            raise FileExistsError('Could not create file \033[1m{}\033[0m'.format(relfname));
    return;

def createFileStamp(
    path: str,
    overwrite: bool,
    options: Dict[str, Any]
):
    if os.path.exists(path) and not overwrite:
        return;
    lines = [];
    border = r'%% ' + '*'*80;
    max_tag_length = max([0] + [len(key) for key in options]);
    for key in options:
        value = options[key];
        tag = key.upper();
        line = r'%% ' + tag + r':';
        if isinstance(value, str):
            value = re.split('\n', str(value));
        elif isinstance(value, (int, float, bool)):
            value = [str(value)];
        if isinstance(value, list) and len(value) == 1:
            line += ' '*(1 + max_tag_length - len(tag)) + str(value[0]);
        elif isinstance(value, list) and len(value) > 1:
            indent = '\n' + r'%% ' + ' '*4;
            line_ = [''];
            line_ += [u for u in value if isinstance(u, str)];
            line += indent.join(line_);
        else:
            line += ' '*(1 + max_tag_length - len(tag)) + r'—';
        lines.append(line);
    if len(lines) > 0:
        lines = [border] + lines + [border];
    writeTextFile(path=path, lines=lines);
    return;

def createFileParameters(
    path: str,
    overwrite: bool,
    options: Dict[str, Any]
):
    if os.path.exists(path) and not overwrite:
        return;
    appconfig.setExportVars({});
    lines = [];
    for key, value in options.items():
        try:
            typ, codedvalue = convertToPythonString(value, indent=0, multiline=False);
            appconfig.setExportVarsKeyValue(key=key, value=value, codedvalue=codedvalue);
            lines.append('<<< global set {key} = {value}; >>>'.format(key = key, value = codedvalue));
        except:
            continue;
    if os.path.isfile(path) and not overwrite:
        return;
    writeTextFile(path=path, lines=lines);
    return;

def createParameters(options: Dict[str, Any]):
    appconfig.setExportVars({});
    lines = [];
    for key, value in options.items():
        try:
            typ, codedvalue = convertToPythonString(value, indent=0, multiline=False);
            appconfig.setExportVarsKeyValue(key=key, value=value, codedvalue=codedvalue);
        except:
            continue;
    return;

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# TERTIARY METHODS
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

def make_file_if_not_exists(path: str, fname: str) -> bool:
    fname_full = os.path.join(path, fname);
    try:
        if not os.path.isfile(fname_full):
            logInfo('File \033[96;1m{}\033[0m will be created.'.format(fname));
            createFile(fname_full);
    except:
        pass;
    return os.path.isfile(fname_full);

def make_dir_if_not_exists(path: str, fname: str) -> bool:
    fname_full = os.path.join(path, fname);
    try:
        if not os.path.isdir(fname_full):
            logInfo('Folder \033[96;1m{}\033[0m will be created.'.format(fname));
            createPath(path=fname_full);
    except:
        pass;
    return os.path.isdir(fname_full);

File: golang/internal/steps/step_create.go
package steps

/* ---------------------------------------------------------------- *
* IMPORTS
* ---------------------------------------------------------------- */

import (
	"fmt"
	"path/filepath"
	"strings"

	"phpytex/internal/core/logging"
	"phpytex/internal/core/utils"
	"phpytex/internal/setup/appconfig"
	"phpytex/internal/types"
)

/* ---------------------------------------------------------------- *
 * METHODS
 * ---------------------------------------------------------------- */

func Create() error {
	var err error = nil
	logging.LogInfo("CREATION STAGE STARTED.")
	var root string = appconfig.Parameters.PathRoot.GetValue()
	createFilesAndFolders(root, appconfig.ProjectTree)
	if appconfig.Parameters.WithFileStamp.GetValue() {
		err = createFileStamp(
			appconfig.Parameters.FileStamp.GetValue(false),
			appconfig.Parameters.OptionOverwriteStamp.GetValue(),
			appconfig.DictionaryStamp.GetValues(),
		)
	}
	if err != nil {
		return err
	}
	if appconfig.Parameters.WithFileParamsPy.GetValue() {
		err = createParameters(appconfig.DictionaryParams.GetValues())
	}
	if err != nil {
		return err
	}
	logging.LogInfo("CREATION STAGE COMPLETE.")
	return nil
}

/* ---------------------------------------------------------------- *
 * SECONDARY METHODS
 * ---------------------------------------------------------------- */

func createFilesAndFolders(path string, projectTree *types.TreeConfig) error {
	if projectTree == nil {
		return nil
	}
	if projectTree.Files != nil {
		for _, fname := range *projectTree.Files {
			err := makeFileIfNotExists(path, fname)
			if err != nil {
				return err
			}
		}
	}
	if projectTree.Folders != nil {
		for relpath, _ := range *projectTree.Folders {
			err := makeDirIfNotExists(path, relpath)
			if err != nil {
				return err
			}
		}
		for relpath, tree := range *projectTree.Folders {
			createFilesAndFolders(filepath.Join(path, relpath), tree)
		}
	}
	return nil
}

func createFileStamp(path string, overwrite bool, options *map[string]interface{}) error {
	if (utils.CheckPathExists(path) && !overwrite) || options == nil {
		return nil
	}

	var (
		err    error
		text   string
		border string
	)

	text, err = utils.DisplayMapAsStamp(*options, "%% ", "  ", false, true, false, false)

	if err != nil {
		return err
	}

	if len(text) > 0 {
		border = "%% " + strings.Repeat(`*`, 80)
		text = strings.Join([]string{border, text, border}, "\n")
	}
	err = utils.WriteTextFile(path, text)

	return err
}

func createFileParameters(path string, overwrite bool, options *map[string]interface{}) error {
	if (utils.CheckPathExists(path) && !overwrite) || options == nil {
		return nil
	}
	var (
		err        error
		lines      []string
		line       string
		text       string
		key        string
		value      interface{}
		codedvalue string
	)

	appconfig.ExportVariables.Init()
	lines = []string{}
	for key, value = range *options {
		codedvalue, err = utils.ConvertToPythonString(value, 0, false, "    ")
		if err != nil {
			break
		}
		appconfig.ExportVariables.SetValue(key, []interface{}{value, codedvalue})
		line = fmt.Sprintf(`<<< global set %[1]s = %[2]v; >>>`, key, codedvalue)
		lines = append(lines, line)
	}
	text = strings.Join(lines, "\n")

	err = utils.WriteTextFile(path, text)

	return err
}

func createParameters(options *map[string]interface{}) error {
	if options == nil {
		return nil
	}
	var (
		err        error
		key        string
		value      interface{}
		codedvalue string
	)

	appconfig.ExportVariables.Init()
	for key, value = range *options {
		codedvalue, err = utils.ConvertToPythonString(value, 0, true, "    ")
		if err != nil {
			break
		}
		appconfig.ExportVariables.SetValue(key, []interface{}{value, codedvalue})
	}

	return err
}

/* ---------------------------------------------------------------- *
 * TERTIARY METHODS
 * ---------------------------------------------------------------- */

func makeFileIfNotExists(path string, fname string) error {
	var err error = nil
	var pathFull string
	pathFull = filepath.Join(path, fname)
	if utils.CheckPathExists(pathFull) {
		if !utils.IsFile(pathFull) {
			return fmt.Errorf("The path \033[96;1m%[1]s\033[0m already exists and cannot be made into a file.", pathFull)
		}
	} else {
		logging.LogInfo(fmt.Sprintf("File \033[96;1m%[1]s\033[0m will be created.", fname))
		err = utils.CreatePath(pathFull)
	}
	return err
}

func makeDirIfNotExists(path string, relpath string) error {
	var err error = nil
	var pathFull string
	pathFull = filepath.Join(path, relpath)
	if utils.CheckPathExists(pathFull) {
		if !utils.IsDir(pathFull) {
			return fmt.Errorf("The path \033[96;1m%[1]s\033[0m already exists and cannot be made into a directory.", pathFull)
		}
	} else {
		logging.LogInfo(fmt.Sprintf("Folder \033[96;1m%[1]s\033[0m will be created.", relpath))
		err = utils.CreatePath(pathFull)
	}
	return err
}
